tests/mock_singleton_helpers.h

```diff
@@ -26,27 +26,39 @@
     {                                                                                                                  \
         return dynamic_cast<mock_class&>(instance());                                                                  \
     }
 
+#define MP_MOCK_SINGLETON_HELPER_TYPES(mock_class, parent_class)                                                       \
+private:                                                                                                               \
+    static constexpr auto academy = [] { return sg::make_scope_guard([] { parent_class::reset(); }); }; /*             \
+                                    use a lambda so that the type gets deduced before decltype below */                \
+                                                                                                                       \
+public:                                                                                                                \
+    using Guard = decltype(academy());                                                                                 \
+    using GuardedMock = std::pair<mock_class*, Guard>;
+
 #define MP_MOCK_SINGLETON_INJECT(mock_class, parent_class)                                                             \
-    [[nodiscard]] static auto inject()                                                                                 \
+public:                                                                                                                \
+    [[nodiscard]] static GuardedMock inject()                                                                          \
     {                                                                                                                  \
         parent_class::reset();                                                                                         \
         parent_class::mock<mock_class>();                                                                              \
-        return std::make_pair(&mock_instance(), sg::make_scope_guard([]() { parent_class::reset(); }));                \
+        return std::make_pair(&mock_instance(), academy());                                                            \
     } // one at a time, please!
 
 #define MP_MOCK_SINGLETON_BOILERPLATE(mock_class, parent_class)                                                        \
+    MP_MOCK_SINGLETON_HELPER_TYPES(mock_class, parent_class)                                                           \
     MP_MOCK_SINGLETON_INSTANCE(mock_class)                                                                             \
     MP_MOCK_SINGLETON_INJECT(mock_class, parent_class)
 
 namespace multipass::test
 {
 
-template <typename ConcreteMock, template <typename MockClass> typename MockCharacter = ::testing::NaggyMock>
+template <typename ConcreteMock, template <typename /*MockClass*/> typename MockCharacter = ::testing::NaggyMock>
```

Comment:
I see the `MockClass` template parameter has been commented out here and elsewhere.  Is having the template parameter still necessary or can it be removed altogether?  I'm not seeing it being used anywhere else but I'm probably missing it somewhere.

Reply:
That inner `template <typename>` is needed for the MockCharacter because Naggy, Nice, and Strict are themselves templates. However, the name of the parameter itself is unused, and the IDE was pointing it out, so I just commented it out.

Reply:
Ah, ok, that makes sense.  Thanks!